Implement ff traits for bls12_381 and jubjub crates

str4d opened the pull request:
Closes #160 and #166.

ebfull commented on bls12_381/Cargo.toml:
```diff
@@ -20,6 +20,15 @@ name = "groups"
 harness = false
 required-features = ["groups"]
 
+[dependencies.ff]
```
I kinda want `ff` /  `group` / `pairing` to be optional dependencies, somehow. I figured `pairing` would be brought in by the `pairings` feature, and that `group` would be brought in by the `groups` feature. I hadn't thought about `ff` though. Hm..

str4d replied:
`fields` feature? 😛

str4d replied:
(and then have `groups =  ["fields"]` which would make it available for both `groups` and `pairings`)

ebfull replied:
That sounds like a good approach, but it means: `fields` enables `Field` but the scalar field implementation is exported anyway, whereas `groups` enables both the `group` APIs and the implementations as well. This is a strange asymmetry.

str4d replied:
Ah, well then `ff` should be a required dependency for the field implementation, given that `group` is a required dependency when using the group implementation.

ebfull replied:
Alright, so we'll make ff a required dependency. You can consider this point resolved.

ebfull commented on bls12_381/src/fp.rs:
```diff
@@ -51,6 +61,27 @@ impl PartialEq for Fp {
 
+impl Ord for Fp {
+    fn cmp(&self, other: &Fp) -> Ordering {
+        // Elements are ordered lexicographically.
```
This isn't constant time; also, although an ordering _can_ be defined for elements of a prime field, I don't think that (strictly speaking) fields should implement `(Partial)Ord`. This sounds like a question for @daira.

str4d replied:
I know there was something non-test-related that required the `Ord` bound, but 1fe3e3784cf9c9f1ec4da0210c6095b56fb3447a appears to only show it needed by a test. I'll try to track down where the usage was.

str4d replied:
Looks like it was only a test that was using this bound; I can remove it.

As a side note, `ff_derive` has implemented `Ord` and `PartialOrd` since https://github.com/zkcrypto/ff/pull/1, where I copied over the API from `pairing::bls12_381` which did at the time implement `(Partial)Ord` on fields. Should these also be removed?

ebfull replied:
If we remove this Ord, we should remove them from derive too yeah.

ebfull commented on bls12_381/src/fp.rs:
```diff
@@ -545,6 +608,181 @@ impl Fp {
+    }
+}
+
+impl BitAnd<u64> for Fp {
```
These (this and the Shr) are weird. I don't think we should support them as part of the `Field` trait's contract. If someone needs to do this stuff they should deal with the existing API somehow.

str4d replied:
They are part of the `PrimeField` trait contract, due to `bellman::multiexp` deriving buckets, and `zcash_primitives::pedersen_hash` generating a window table. This is one of the constraints I was contemplating moving to a `ScalarField` trait, so that we could implement these only for `bls12_381::Scalar` and `jubjub::Fr`, not `bls12_381::Fp`.

str4d replied:
See 08500ee71275e7b6127e7b512c0763b058acf714 and 49f119fb033dac88165028b7117e3d23e9f469a8 (the `bellman::multiexp` change was accidentally left in the latter).

ebfull replied:
Yeah, those two examples are niche. I think we can implement this inline and not provide it as part of the crate. I also think the implementation of multiexp is weird and shouldn't need to use those operations... well, probably anyway.

ebfull commented on bls12_381/src/fp.rs:
```diff
@@ -545,6 +608,181 @@ impl Fp {
+
+impl Field for Fp {
+    fn random<R: RngCore + ?Sized>(_rng: &mut R) -> Self {
+        unimplemented!()
```
Not implemented.

str4d replied:
Do we need to be able to generate random base field elements? Or should we instead move this to a `ScalarField` trait?

ebfull replied:
I don't think we need to, but it could be reasonable that someone downstream would.

str4d replied:
I'm going to implement this as a 768-bit reduction, matching the 2-to-1 512-bit reduction in `Scalar::from_bytes_wide`, unless @daira says otherwise.

ebfull commented on bls12_381/src/fp.rs:
```diff
@@ -545,6 +608,181 @@ impl Fp {
+        self.invert()
+    }
+
+    fn frobenius_map(&mut self, _power: usize) {
```
I don't think we need to offer this as part of the `Field` trait since it's only ever used internally in `bls12_381`, and field extensions aren't exposed anywhere in our stack.

str4d replied:
Cool, we can drop this.

ebfull commented on bls12_381/src/fp.rs:
```diff
@@ -545,6 +608,181 @@ impl Fp {
+    }
+}
+
+impl SqrtField for Fp {
```
I want to get rid of this trait.

str4d replied:
Should `sqrt` then be exposed on `PrimeField` or `Field` (with the `CtOption` used to indicate both "this element has no square root" and "this field does not have a square root operation"), or not be accessible generically?

ebfull replied:
`Field`. All fields should be able to support square roots.

str4d commented:
I've addressed most comments in #228. Once that PR has been reviewed and merged, I'll rebase this PR to remove the now-unnecessary parts, and implement `Field::random`.

str4d commented:
Rebased on master after merging #228.

str4d commented:
The first commit in this PR is also in #230; I'll rebase whichever PR is not merged first.

str4d commented:
Rebased on master now that #230 has been merged.

str4d commented:
Rebased the PR; it now only implements the `ff` traits for the structs that are used as scalars in the group implementations, because the `Base` associated types are being removed from the `group` traits.

daira commented on jubjub/src/fr.rs:
```diff
@@ -70,6 +78,33 @@ pub const MODULUS: Fr = Fr([
     0x0e7d_b4ea_6533_afa9,
 ]);
 
+const MODULUS_BYTES: [u8; 32] = [
```
Say what byte order this is in (maybe include it in the name).

daira replied:
Also consider renaming `Fr::from_bytes` to `from_le_bytes`.

ebfull replied:
See the other comment; the same holds here.

ebfull commented on bls12_381/src/scalar.rs:
```diff
@@ -70,6 +78,22 @@ const MODULUS: Scalar = Scalar([
     0x73ed_a753_299d_7d48,
 ]);
 
+const MODULUS_BYTES: [u8; 32] = [
```
Note that this value is only used later in the `PrimeField` impl, where the constant is defined to be little endian by the implementation, so this property is documented and accounted for.

str4d commented on bls12_381/src/fp.rs:
```diff
@@ -162,7 +162,7 @@ impl Fp {
-    /// Attempts to convert a little-endian byte representation of
+    /// Attempts to convert a big-endian byte representation of
     /// a scalar into an `Fp`, failing if the input is not canonical.
     pub fn from_bytes(bytes: &[u8; 48]) -> CtOption<Fp> {
```
I'd personally rather not; it implies a little-endian scalar encoding would also be correct and supported (rather than big-endian being canonical for `Fp`), and that way lies madness.
